internal/testutils/proxy.go

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package testutils
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+)
+
+const defaultTestTimeout = 10 * time.Second
+
+// ProxyServer is a proxy server that is used for testing.
+type ProxyServer struct {
+	lis net.Listener
+	in  net.Conn
+	out net.Conn
+
+	requestCheck func(*http.Request) error
+}
+
+// Stop functions stops and cleans up the proxy server.
+func (p *ProxyServer) Stop() {
+	p.lis.Close()
+	if p.in != nil {
+		p.in.Close()
+	}
+	if p.out != nil {
+		p.out.Close()
+	}
+}
+
+// NewProxyServer create and starts a proxy server.
+func NewProxyServer(lis net.Listener, requestCheck func(*http.Request) error, errCh chan error, doneCh chan struct{}, backendAddr string, resolutionOnClient bool, proxyServerStarted func()) *ProxyServer {
```

Comment:
may be we can do something like this fake server in its own package https://github.com/grpc/grpc-go/blob/17d08f746b51fa70911754d36a799c247a2c4b79/internal/testutils/xds/fakeserver/server.go#L19 ? wdyt?

New should only create the object and StartServer/Run should start the go routine to accept requests